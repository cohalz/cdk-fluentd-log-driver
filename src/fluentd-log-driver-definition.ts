import { CfnTaskDefinition, LogDriver } from "@aws-cdk/aws-ecs"
import { Construct } from "@aws-cdk/cdk"

/**
 * Properties for defining a new Fluentd Log Driver
 */

export interface FluentdLogDriverProps {
  readonly env?: string
  readonly envRegex?: string
  readonly fluentdAddress?: string
  readonly fluentdAsyncConnect?: string
  readonly fluentdBufferLimit?: string
  readonly fluentdMaxRetries?: string
  readonly fluentdRetryWait?: string
  readonly fluentdSubSecondPrecision?: string
  readonly labels?: string
  readonly tag?: string
}

/**
 * A log driver that will log to the Fluentd collector
 */
export class FluentdLogDriver extends LogDriver {
  constructor(
    scope: Construct,
    id: string,
    private readonly props?: FluentdLogDriverProps
  ) {
    super(scope, id)
  }

  /**
   * Called when the log driver is configured on a container
   */
  public bind(): void {
    return
  }

  /**
   * Return the log driver CloudFormation JSON
   */
  public renderLogDriver(): CfnTaskDefinition.LogConfigurationProperty {
    if (this.props === undefined) {
      return {
        logDriver: "fluentd",
      }
    }

    const options = removeEmpty({
      env: this.props.labels,
      "env-regex": this.props.envRegex,
      "fluentd-address": this.props.fluentdAddress,
      "fluentd-async-connect": this.props.fluentdAsyncConnect,
      "fluentd-buffer-limit": this.props.fluentdBufferLimit,
      "fluentd-max-retries": this.props.fluentdMaxRetries,
      "fluentd-retry-wait": this.props.fluentdRetryWait,
      "fluentd-sub-second-precision": this.props.fluentdSubSecondPrecision,
      labels: this.props.labels,
      tag: this.props.tag,
    })

    return {
      logDriver: "fluentd",
      options,
    }
  }
}

/**
 * Remove undefined values from a dictionary
 *
 * See: https://github.com/awslabs/aws-cdk/blob/v0.31.0/packages/@aws-cdk/aws-ecs/lib/log-drivers/aws-log-driver.ts#L89-L99
 */
function removeEmpty<T>(x: {
  [key: string]: T | undefined
}): { [key: string]: T } {
  for (const key of Object.keys(x)) {
    if (!x[key]) {
      delete x[key]
    }
  }
  return x as any
}
