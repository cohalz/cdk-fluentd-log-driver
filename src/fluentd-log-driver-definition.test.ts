import { SynthUtils } from "@aws-cdk/assert"
import { ContainerImage, Ec2TaskDefinition } from "@aws-cdk/aws-ecs"
import { Stack } from "@aws-cdk/core"
import { FluentdLogDriver } from "./fluentd-log-driver-definition"

describe("FluentdLogDriverDefinition", () => {
  test("default", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition")

    const logDriver = new FluentdLogDriver()

    taskDefinition.addContainer("Container", {
      image: ContainerImage.fromRegistry("/aws/aws-example-app"),
      logging: logDriver,
      memoryLimitMiB: 2048,
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })

  test("can set fluentd-address", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition")

    const logDriver = new FluentdLogDriver({
      fluentdAddress: "127.0.0.1:24224",
    })

    taskDefinition.addContainer("Container", {
      image: ContainerImage.fromRegistry("/aws/aws-example-app"),
      logging: logDriver,
      memoryLimitMiB: 2048,
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })

  test("can set all options", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition")

    const logDriver = new FluentdLogDriver({
      env: "env",
      envRegex: "^(os|customer).",
      fluentdAddress: "127.0.0.1:24224",
      fluentdAsyncConnect: "true",
      fluentdBufferLimit: "8MB",
      fluentdMaxRetries: "20",
      fluentdRetryWait: "2",
      fluentdSubSecondPrecision: "true",
      labels: "labels",
      tag: "tag",
    })

    taskDefinition.addContainer("Container", {
      image: ContainerImage.fromRegistry("/aws/aws-example-app"),
      logging: logDriver,
      memoryLimitMiB: 2048,
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})
