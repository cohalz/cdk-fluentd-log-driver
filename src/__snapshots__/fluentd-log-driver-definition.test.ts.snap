// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluentdLogDriverDefinition can set all options 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Essential": true,
            "Image": "/aws/aws-example-app",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "LogConfiguration": Object {
              "LogDriver": "fluentd",
              "Options": Object {
                "env": "labels",
                "env-regex": "^(os|customer).",
                "fluentd-address": "127.0.0.1:24224",
                "fluentd-async-connect": "true",
                "fluentd-buffer-limit": "8MB",
                "fluentd-max-retries": "20",
                "fluentd-retry-wait": "2",
                "fluentd-sub-second-precision": "true",
                "labels": "labels",
                "tag": "tag",
              },
            },
            "Memory": 2048,
            "MountPoints": Array [],
            "Name": "Container",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ecs-tasks.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`FluentdLogDriverDefinition can set fluentd-address 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Essential": true,
            "Image": "/aws/aws-example-app",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "LogConfiguration": Object {
              "LogDriver": "fluentd",
              "Options": Object {
                "fluentd-address": "127.0.0.1:24224",
              },
            },
            "Memory": 2048,
            "MountPoints": Array [],
            "Name": "Container",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ecs-tasks.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`FluentdLogDriverDefinition default 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Essential": true,
            "Image": "/aws/aws-example-app",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "LogConfiguration": Object {
              "LogDriver": "fluentd",
            },
            "Memory": 2048,
            "MountPoints": Array [],
            "Name": "Container",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ecs-tasks.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
